dask/order.py

```diff
@@ -274,177 +276,467 @@ def finish_now_key(x):
+                        break
+                later_singles_clear()
+                deps = set_difference(deps, result)
+                if not deps:
                     continue
+                add_to_inner_stack = False
+                process_singles = True
+            else:
+                break
 
-                result[item] = i
-                i += 1
-                deps = dependents[item]
-
-                # If inner_stack is empty, then we typically add the best dependent to it.
-                # However, we don't add to it if we complete a node early via "finish_now" below
-                # or if a dependent is already on an inner_stack.  In this case, we add the
-                # dependents (not in an inner_stack) to next_nodes or later_nodes to handle later.
-                # This serves three purposes:
-                #   1. shrink `deps` so that it can be processed faster,
-                #   2. make sure we don't process the same dependency repeatedly, and
-                #   3. make sure we don't accidentally continue down an expensive-to-compute path.
-                add_to_inner_stack = True
-                if metrics[item][3] == 1:  # min_height
-                    # Don't leave any dangling single nodes!  Finish all dependents that are
-                    # ready and are also root nodes.
-                    finish_now = {
-                        dep
-                        for dep in deps
-                        if not dependents[dep] and num_needed[dep] == 1
-                    }
-                    if finish_now:
-                        deps -= finish_now  # Safe to mutate
-                        if len(finish_now) > 1:
-                            finish_now = sorted(finish_now, key=finish_now_key)
-                        for dep in finish_now:
-                            result[dep] = i
-                            i += 1
-                        add_to_inner_stack = False
-
-                if deps:
-                    for dep in deps:
-                        num_needed[dep] -= 1
-
-                    already_seen = deps & seen
-                    if already_seen:
-                        if len(deps) == len(already_seen):
-                            continue
-                        add_to_inner_stack = False
-                        deps -= already_seen
-
-                    if len(deps) == 1:
-                        # Fast path!  We trim down `deps` above hoping to reach here.
-                        (dep,) = deps
-                        if not inner_stack:
+            if process_singles and singles:
+                # We gather all dependents of all singles into `deps`, which we then process below.
+                # A lingering question is: what should we use for `item`?  `item_key` is used to
+                # determine whether each dependent goes to `next_nodes` or `later_nodes`.  Currently,
+                # we use the last value of `item` (i.e., we don't do anything).
```

Comment:
For posterity, this is the main question in this PR that I'm not quite satisfied with.  We sometimes compare the dependent partition_keys to the partition_key of the current item, but when we process the dependents of `singles`, we don't have a good `item` to use.

In practice, it doesn't seem to matter what we use here.  I experimented setting item to a faux `min_item` that has `partition_key[min_item] == -1`, which causes dependents to go to `later_nodes` instead of `next_nodes`.  Again, I couldn't find anywhere where this actually made a difference.  Another option might be to set `item` to the item on the top of the inner stack.